Migrate from `open-color` to `colar`

kelvindecosta opened the pull request:
## Overview

This PR attempts to migrate from using `open-color` to `colar`.
The latter is based on the former and extends the color palette.

![Screenshot_20221116_220142](https://user-images.githubusercontent.com/38982451/202269800-02a49d58-7e19-4724-a08b-64659e663d2c.png)

## Notes

Here are some changes that I thought were worth highlighting:

- Replace `--grape` with `--purple` (palette values are unchanged)
- Add `colorjs.io`, required by `props.colors.src.js`, as a `devDependency` as its CDN seems to be unavailable
- Set `.color-swatch` `inline-size` as ~~`7%`~~ `6.7%`, to account for `13` luminosity steps.

I decided not to make changes to any theming (including the documentation).
It could use some love:

![Screenshot_20221116_220402](https://user-images.githubusercontent.com/38982451/202270223-35beb2d7-1fe1-46cb-af66-a7f43804610e.png)

![Screenshot_20221116_220428](https://user-images.githubusercontent.com/38982451/202270307-81475f86-40d2-4a5a-8afa-94946c494edd.png)

## Related Issues

Should fully/partially address #247 & #308

## Conclusion

I tried to be as systematic and careful as possible.

Any feedback is appreciated!

kelvindecosta commented:
Hey

Not sure why there are so many commits. I'll try to get this fixed ASAP

argyleink commented:
this looks awesome. did a visual review of the code, nice work in there being minimal but getting the work done, ty 🙏🏻 
i'll pull your work down onto a local machine soon and give it a peep and get back to you 🙂

argyleink commented:
the dark side of the grays are really muddy don't you think? didnt notice that when reviewing colar before. we may need to intervene and supply our own. i took a quick stab at updating the normalize surfaces and text and it's just all so brown feeling. the light theme is mostly unaffected, except the dark text colors look a little muddy too. 

<img width="1252" alt="Screen Shot 2022-11-16 at 2 22 30 PM" src="https://user-images.githubusercontent.com/1134620/202307332-70fb8984-4c32-46e0-a692-5d3409b8e8fa.png">


also, the darkest variants dont look perceptually equal, like orange 12 vs sand 12.

that was my 5m review hehe, will dig in and provide more specific feedback later. it's definitely a step forward!

argyleink commented:
so it looks like just the darker grays that go warm in colar, why the dark theme was so noticable, you can even see the hue tilt in the angle of hsl:

![Screen Shot 2022-11-16 at 8 18 24 PM](https://user-images.githubusercontent.com/1134620/202354202-db222eee-8a9c-429b-ad3d-3eaa35cecd8c.png)

we could keep the colar set and just call it warm gray, then keep using the open color grays as a base? this would mean all the current normalize uses would look the exact same, but there'd be 3 new darker colors available. i've been making darker grays for various reasons and have some ready to go. i plugged them into a favorite tool of mine and tweaked them a bit, then tried them out in the site.

`--gray-10` `#16191d`
`--gray-11` `#0d0f12`
`--gray-12` `#030507`

i need to test them on some different screens, make sure they're not over saturated. might also try doing the adjustment in oklch instead of hsl, see if i can get a little bit more of a steady color ramp. thoughts on appending 3 colors to open-color like this, keeping the gray set like it was, and adding colar as a warm set?

![Screen Shot 2022-11-16 at 8 43 55 PM](https://user-images.githubusercontent.com/1134620/202358316-8637ffc6-5fe8-41ac-8507-c7ed9ac7bfd7.png)

![Screen Shot 2022-11-16 at 8 45 04 PM](https://user-images.githubusercontent.com/1134620/202358319-4daaea31-7bba-496a-a6db-ab1fdf398340.png)

what's cool here at first, was that the dark theme looked the same as it did before. but i updated the surfaces to use some of the darker variants i show above, and it's nice i felt:

![Screen Shot 2022-11-16 at 8 51 47 PM](https://user-images.githubusercontent.com/1134620/202359275-05c85b8f-1c7f-4582-b42e-353586e50a9d.png)

todo:
- [ ] update docsite gradients from what was something like `--indigo-5` to the newer bright swatch in colar<br>likely `--indigo-6` but maybe 7. will need an eye, as i noticed because some of the gradient had lost some vibrancy on the docsite.

argyleink reviewed (changes requested):
very cool work in here, just gotta say again thanks for taking the time and making the most minimal changes in here.

i think we should hack on the grays and not break dark themes. i'll do some more testing across screens and triple check the export fields. PR needs to update some [test counts](https://github.com/argyleink/open-props/blob/main/test/basic.test.cjs#L7) to pass the unit tests. then we're probably g2g 🙂

kelvindecosta commented:
Hey Adam

Thank you for the detailed feedback!

> the dark side of the grays are really muddy don't you think? didnt notice that when reviewing colar before. we may need to intervene and supply our own. i took a quick stab at updating the normalize surfaces and text and it's just all so brown feeling. the light theme is mostly unaffected, except the dark text colors look a little muddy too.

Yes, it feels much warmer than a neutral gray. 
Not a fan of Tailwind's class-based approach, but they actually have a nice palette of grays.
https://tailwindcss.com/docs/customizing-colors

> we could keep the colar set and just call it warm gray, then keep using the open color grays as a base? this would mean all the current normalize uses would look the exact same, but there'd be 3 new darker colors available. i've been making darker grays for various reasons and have some ready to go. i plugged them into a favorite tool of mine and tweaked them a bit, then tried them out in the site.
> i need to test them on some different screens, make sure they're not over saturated. might also try doing the adjustment in oklch instead of hsl, see if i can get a little bit more of a steady color ramp. thoughts on appending 3 colors to open-color like this, keeping the gray set like it was, and adding colar as a warm set?

I like this solution. Just to be clear, the final palette would be something like:

- `colar.gray` -> `open-props.warm-gray` (or a one-word name for `warm-gray` to stick with the pattern)
- `open-color.gray` + `open-props.gray{10-12}` -> `open-props.gray`  
- everything else from `colar`

Giving developers the choice of a warmer gray is really nice.
Do you think `open-props` should provide an option for a cooler gray, with its hue tending towards blue, similar to Tailwind's `slate` ?
![image](https://user-images.githubusercontent.com/38982451/202534603-cd262f23-8858-4d2f-aa5c-94806dd3e97c.png)

> what's cool here at first, was that the dark theme looked the same as it did before. but i updated the surfaces to use some of the darker variants i show above, and it's nice i felt:

The new surfaces look really clean!

---

I have a few questions:
- How are the `.min.css` files in the `docsite` folder generated? Are they among the output files of  `npm build`?
- What tool(s) do you use to study colors? I really like the a11y ratio chart. 

I'll try to make the following changes over the coming weekend:

- updating tests to account for new colors
- accounting for gray from `open-color` and warm gray from `colar`

If there's anything else you'd want me to try out, please let me know.

Thank you for your time!

argyleink commented:
just tested the 3 new dark grays in a variety of situations and screens, they're g2g

argyleink commented:
> they actually have a nice palette of grays

they do, agree, but it's also too many options maybe? we can def get away with a warm and a cool set. i do like their name of "stone"

> colar.gray -> open-props.warm-gray (or a one-word name for warm-gray to stick with the pattern)
> open-color.gray + open-props.gray{10-12} -> open-props.gray
> everything else from colar

yep! what do you think of calling colar's gray palette "stone" as the one word name?

> How are the .min.css files in the docsite folder generated? Are they among the output files of npm build?

they jump up a directory and grab open props from src like. and i think this answers the question?!

```css
@import "../src/index.css";
@import "../src/extra/normalize.css";
```

> What tool(s) do you use to study colors? I really like the a11y ratio chart.

i dont have a favorite, but for this task i used https://components.ai/theme

<br>
thanks again, lovin this progress. i'm excited to merge it and update a couple things 💯

kelvindecosta commented:
> they do, agree, but it's also too many options maybe? we can def get away with a warm and a cool set. i do like their name of "stone"
> yep! what do you think of calling colar's gray palette "stone" as the one word name?

`stone` is nice!

---

> > How are the .min.css files in the docsite folder generated? Are they among the output files of npm build?
> they jump up a directory and grab open props from src like. and i think this answers the question?!
> ```css
> @import "../src/index.css";
> @import "../src/extra/normalize.css";
> ```

Ah yes, that makes sense for the `.css` files that aren't minified.

Are the files in `docsite/public` generated by the `lib:*` scripts?
I only just noticed that their paths match the path supplied via the `-o` option in those commands.

```
docsite
└── public
    ├── media.min.css # npm:lib:media
    ├── normalize.min.css # npm:lib:normalize
    └── open-props.min.css # npm:lib:all
```

If this is the case, should I update them to reflect the new palette?

---

> i dont have a favorite, but for this task i used https://components.ai/theme

Wow, this is really awesome, I really like their interface!
Thank you for sharing it!

kelvindecosta commented:
## Updates

The palette is updated as follows:

```js
// Combine `open-color` and `colar` palettes
const colors = Object.entries({
  // Extend `openColor.gray`
  gray: {
    ...openColor.gray,
    10: "#16191d",
    11: "#0d0f12",
    12: "#030507",
  },
  // Use `colar.gray` as "stone"
  stone: colar.gray,
  // Use all other colors in `colar`
  ...Object.fromEntries(
    Object.entries(colar).filter(([hueName]) => hueName != "gray")
  ),
});
```

Also, updates were made to:

- generated `props.[color].css` files
- package exports and scripts
- `Colors` section in documentation

## Tests

The changes so far, update the palette by:

- Adding 3 shades for each existing color, (13 colors in `open-color`) `[3 * 13 = 39]`
- Add 6 new hues (`stone`, `choco`, `brown`, `sand`, `camo`, & `jungle`) each with 13 shades `[6 * 13 = 78]`

In total, there are `117` new shades, or `open-props` tokens.

```diff
@@ -24,7 +24,7 @@ test('Import should have colors', async t => {
 })

 test('JSON Import should have colors', async t => {
-  t.is(Object.keys(OPtokens).length, 280)
+  t.is(Object.keys(OPtokens).length, 397)
   t.assert(Object.keys(OPtokens).includes('--orange-0'))
 })
```

Each color token contributes 4 to this count, so a difference of `[3 * 117 = 468]`:

```diff
@@ -4,7 +4,7 @@ const OpenProps = require('../dist/open-props.cjs')
 const OPtokens = require('../open-props.tokens.json')

 test('Should have an all included import', t => {
-  t.is(Object.keys(OpenProps).length, 1098)
+  t.is(Object.keys(OpenProps).length, 1566)
 })
```

## Notes

- > update docsite gradients from what was something like --indigo-5 to the newer bright swatch in colar
  > likely --indigo-6 but maybe 7. will need an eye, as i noticed because some of the gradient had lost some vibrancy on the docsite.

  I assumed you'd want to do this together with the updates to `--surface` and other derived color props.

- I think `docsite/public/*.min.css` files might need an update
## Updates

The palette is updated as follows:

```js
// Combine `open-color` and `colar` palettes
const colors = Object.entries({
  // Extend `openColor.gray`
  gray: {
    ...openColor.gray,
    10: "#16191d",
    11: "#0d0f12",
    12: "#030507",
  },
  // Use `colar.gray` as "stone"
  stone: colar.gray,
  // Use all other colors in `colar`
  ...Object.fromEntries(
    Object.entries(colar).filter(([hueName]) => hueName != "gray")
  ),
});
```

Also, updates were made to:

- generated `props.[color].css` files
- package exports and scripts
- `Colors` section in documentation

## Tests

The changes so far, update the palette by:

- Adding 3 shades for each existing color, (13 colors in `open-color`) `[3 * 13 = 39]`
- Add 6 new hues (`stone`, `choco`, `brown`, `sand`, `camo`, & `jungle`) each with 13 shades `[6 * 13 = 78]`

In total, there are `117` new shades, or `open-props` tokens.

```diff
@@ -24,7 +24,7 @@ test('Import should have colors', async t => {
 })

 test('JSON Import should have colors', async t => {
-  t.is(Object.keys(OPtokens).length, 280)
+  t.is(Object.keys(OPtokens).length, 397)
   t.assert(Object.keys(OPtokens).includes('--orange-0'))
 })
```

Each color token contributes 4 to this count, so a difference of `[3 * 117 = 468]`:

```diff
@@ -4,7 +4,7 @@ const OpenProps = require('../dist/open-props.cjs')
 const OPtokens = require('../open-props.tokens.json')

 test('Should have an all included import', t => {
-  t.is(Object.keys(OpenProps).length, 1098)
+  t.is(Object.keys(OpenProps).length, 1566)
 })
```

## Notes

- > update docsite gradients from what was something like --indigo-5 to the newer bright swatch in colar
  > likely --indigo-6 but maybe 7. will need an eye, as i noticed because some of the gradient had lost some vibrancy on the docsite.

  I assumed you'd want to do this together with the updates to `--surface` and other derived color props.

- I think `docsite/public/*.min.css` files might need an update

argyleink commented on docsite/index.html:
```diff
@@ -1469,18 +1610,18 @@ <h5>Dim</h5>
             </div>
             <div>
-              <h5>Grape</h5>
+              <h5>Jungle</h5>
```
```suggestion
              <h5>Purple</h5>
```

argyleink commented on docsite/index.html:
```diff
@@ -1469,18 +1610,18 @@ <h5>Dim</h5>
               <pre class="language-css"><code>
                 html {
-                  --brand-grape: var(--grape-5);
+                  --brand-jungle: var(--jungle-5);
```
```suggestion
                  --brand-purple: var(--purple-5);
```

argyleink commented on docsite/index.html:
```diff
@@ -1469,18 +1610,18 @@ <h5>Dim</h5>
 
-                  --text1-grape: var(--grape-9);
-                  --text2-grape: var(--grape-7);
+                  --text1-jungle: var(--jungle-9);
+                  --text2-jungle: var(--jungle-7);
```
```suggestion
                  --text1-purple: var(--purple-9);
                  --text2-purple: var(--purple-7);
```

argyleink commented on docsite/index.html:
```diff
@@ -1469,18 +1610,18 @@ <h5>Dim</h5>
-                  --surface2-grape: var(--grape-1);
-                  --surface3-grape: var(--grape-2);
-                  --surface4-grape: var(--grape-3);
+                  --surface1-jungle: var(--jungle-0);
+                  --surface2-jungle: var(--jungle-1);
+                  --surface3-jungle: var(--jungle-2);
+                  --surface4-jungle: var(--jungle-3);
```
```suggestion
                  --surface1-purple: var(--purple-0);
                  --surface2-purple: var(--purple-1);
                  --surface3-purple: var(--purple-2);
                  --surface4-purple: var(--purple-3);
```

argyleink commented on docsite/index.html:
```diff
@@ -1546,18 +1687,18 @@ <h5>Prefers Dim</h5>
               <pre class="language-css"><code>
-                [color-scheme="grape"] {
-                  --brand: var(--brand-grape);
+                [color-scheme="jungle"] {
+                  --brand: var(--brand-jungle);
```
```suggestion
                [color-scheme="purple"] {
                  --brand: var(--brand-purple);
```

argyleink reviewed (changes requested):
i'm ready to approve, just given the small suggestions here are accepted 👍🏻 

i [removed the legacy compiled files in public](https://github.com/argyleink/open-props/commit/5d053a6ac4e16291e541b288560eabed0068428f), pull latest and you wont need to worry about them. 

is it easy for you to change your branch target for this pr to `beta`? if so, can you swap that so I can do wrap up work in a subsequent PR to the beta branch? i've got a todo list of tasks once these new colors merge in and i might not be able to get to it same day. if it's not easy, no worries!

kelvindecosta commented:
Hey Adam,

Short summary of the changes:

- Switched the target branch to `beta`
- Applied suggested changes from the code review
- Removed changes to `.min.css` files

> I messed something up while pulling the changes from the `main` branch and running `git rebase main` from the local `colar` branch.
> For some reason, there are 9 more commits than necessary, but the net changes are not affected.
> If you find some issue please do let me know!

Hopefully, we're good to go!

argyleink commented:
> the net changes are not affected

agreed, looks fine, and i'm gonna squash it anyway 👍🏻
